lib/Doctrine/ORM/Internal/Hydration/SimpleObjectHydrator.php

```diff
@@ -129,6 +132,11 @@ protected function hydrateRowData(array $sqlResult, array &$result)
 
             // Prevent overwrite in case of inherit classes using same property name (See AbstractHydrator)
             if (! isset($data[$fieldName]) || ! $valueIsNull) {
+                // If we have inheritance in resultset, make sure the field belongs to the correct class
+                if (isset($cacheKeyInfo['discriminatorValues']) && ! in_array($discrColumnValue, $cacheKeyInfo['discriminatorValues'])) {
```

Comment:
It would be safe to perform a strict `in_array()` here, right? If so we need to add `true` as third argument, otherwise ensure that non-strict comparison here is tested.